Feature #1129 multi study context

jan-petr opened the pull request:
### Linked issue

Closes #1129

### How to test

Check flavor /Siemens_PCASL_3DGRASE_VD13A_DEBBIE_1 and Siemens_PCASL_multi_VE11C_1

HenkMutsaerts commented on Functions/InputOutput/xASL_io_ReadDataPar.m:
```diff
@@ -147,37 +147,56 @@
-				if exist(x.(sFields{n}).(loadPaths{m}),'file')
-					load(fullfile(x.(sFields{n}).(loadPaths{m})));
+		sFields = fieldnames(subx);
+		for n=1:size(sFields,1)
```
please let's not use single-letter variables. subx m n should all be readable words

jan-petr replied:
subx = sub-x. x is already a one letter variable.
m and n replaced.

HenkMutsaerts replied:
OK then please call it `subjectX` and/or clarify in the header what subx means

jan-petr replied:
`subfieldX` - clarified in the comment, not in the header, as this is only a local variable.

HenkMutsaerts replied:
Great, good example that abbreviating costs time. I would have needed to figure out that `sub` didn't refer to `subject` but to `subfield`. Writing this out saves time.

Please resolve.

HenkMutsaerts commented on Functions/InputOutput/xASL_io_ReadDataPar.m:
```diff
@@ -147,37 +147,56 @@
 		end
+		
+		% Convert strings containing numbers to number
+		subx = xASL_io_ReadDataPar_FixFields(subx);
```
note that this is not on subject level but on SubjVisitSess level

jan-petr replied:
Not sure if I understand. This is once per studyPar = once per the whole import.

HenkMutsaerts replied:
ah OK I thought studyPar is now repeated for each possible combination. So for each subject/visit/session level?

jan-petr replied:
this function checks and cleans studyPar (saved in x). If there are multiple studyPars, then we go one by one, save each to `subfieldX`, clean it using the normal procedure, and then put it back cleaned to the list of studyPars.

We don't repeat for all combinations of subject/visit/session, we simply go through the list of studyPars.

HenkMutsaerts commented on Modules/SubModule_Import/xASL_wrp_NII2BIDS.m:
```diff
@@ -48,18 +48,28 @@
 	
 	%% 2. Create the study description output and verify that all is there
-	datasetDescription = xASL_bids_CreateDatasetDescriptionTemplate(studyPar, x.Version);
+	datasetDescription = xASL_bids_CreateDatasetDescriptionTemplate(studyParFirstContext, x.Version);
```
What if there are "more exceptions than the rule"? Why should the first be the one to use for the DataSetDescription? Probably the average of all is better...

jan-petr replied:
Average is impossible to do. The basic idea is that the first (lowest priority) entry is the one that is common for all and contains the basics, and all the others are additional things specific only in special cases. We should explain that better in the documentation. But it doesn't make sense to make an average. Worst case - we try to see if there are extra-entries relevant for datasetDescription - but the point is they shouldn't.

HenkMutsaerts replied:
This is a dangerous assumption; assuming that there is a basic protocol and that exceptions are rare. We observe in multiple studies (OASIS, NOMARED, EPAD, etc etc) that this is not the case.

So example case:
EPAD has 1600 subjects, 15 different sites. Which site is the "first" or "common" entry? (note that all this needs to be multi-sequence multi-vendor etc ready for Mathijs and ABBA)

jan-petr replied:
I don't have the answer - but this has nothing to do with my code...
The information to be written to datasetDescription has to be decided by study authors, it cannot be compiled from info inside the studyPar automatically. They can't write different datasetDescription tags to for each sequence and expect that this is merged automatically.

If you want, I can forbid to do overwriting for the datasetDescription-tags. But this has nothing to do with my JSON - that would have been the same problem for the table of course.

Besides - datasetDescription is about author, acknowledgment etc - it has nothing to do with the protocol per site.

HenkMutsaerts replied:
OK now I understand. This is solved by adding a warning for each inheritance case.

HenkMutsaerts commented on Modules/SubModule_Import/xASL_wrp_NII2BIDS_Subject.m:
```diff
@@ -1,13 +1,13 @@
 %   x.modules.import.imPar - JSON file with structure with import parameter (REQUIRED, STRUCT)
 %   bidsPar                - Output of xASL_imp_Config (REQUIRED, STRUCT)
-%   studyPar               - JSON file with the BIDS parameters relevant for the whole study (REQUIRED, STRUCT)
+%   studyParFull           - JSON file with the BIDS parameters relevant for the whole study, potentially multi-context (REQUIRED, STRUCT)
```
"potentiall multi-context" -> "supports providing nested studyPars for all potential subject/session/run permutations/combinations"

jan-petr replied:
fixed

HenkMutsaerts commented on Modules/SubModule_Import/xASL_wrp_NII2BIDS_Subject.m:
```diff
@@ -46,7 +46,17 @@
+		runName = listRuns{iRun};
+		iName = find(runName=='_');
+		if isempty(iName)
+			runName = '';
```
first run name should always default to "run-1"to avoid conflicts and to easy programming

jan-petr replied:
we've discussed this and there's an issue for that...

HenkMutsaerts replied:
Sure, but now you make the new issue larger ;) Then this occassion should perhaps be added to this new issue?

jan-petr replied:
This is the same line/file you talked about previously. There is no other instance to change the code but this one. The issue stays the same.

HenkMutsaerts reviewed (changes requested):
Very nice! Few points

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_Subject_SortASLVolumes.m:
```diff
@@ -165,19 +165,25 @@
-            end
+            % Get the specific studyPar context
+			studyParFull = xASL_io_ReadDataPar(x.dir.studyPar, true);
+			structSubject = x.overview.(['subject_' num2str(iSubject,'%.3d')]);
```
`overview` is indeed a new thing created by Michael for the import - and a necessary one.
But yes, we can rename it to `x.importOverview` = done.

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_Subject_SortASLVolumes.m:
```diff
@@ -165,19 +165,25 @@
 
 %% Determine if we have a Hadamard sequence based on the parameters of the studyPar.json
-function [bTimeEncoded] = xASL_imp_DCM2NII_CheckIfTimeEncoded(x, bTimeEncoded)
+function [bTimeEncoded] = xASL_imp_DCM2NII_CheckIfTimeEncoded(x, bTimeEncoded, iSubject, iVisit, iSession)
```
Part one of your question - it is specific to Hadamard. We are figuring out if this is hadamard or isn't, because we need that information for Hadamard. But we don't need to know in the import if a sequence is multi-TE or multi-PLD. So that's why we need this function and don't need the others.

Part two - I don't understand that part of the question. It doesn't do an import overview. It is just trying to figure out if a specific subject+visit+session+run is or isn't hadamard, because then the import is done a bit differently.

Overall - this is all fine to me.

HenkMutsaerts replied:
It was defining general subject/visit/session fields, that you need in multiple import related functions, not only in the Hadamard function. See the structs below, these definition are needed throughout the import, not only for Hadamard.

>>	studyParAll = xASL_io_ReadDataPar(x.dir.studyPar, true);
	structSubject = x.importOverview.(['subject_' num2str(iSubject,'%.3d')]);
	structVisit   = structSubject.(['visit_' num2str(iVisit,'%.3d')]);
	structRun     = structVisit.(['run_' num2str(iSession,'%.3d')]);
	studyParSpecific = xASL_imp_StudyParPriority(studyParAll, structSubject.name, structVisit.name, structRun.regexp(2:end));

jan-petr replied:
It is a very special case where you need to know in the DCM2NII step if a scan is hadamard or not. you don't need this anywhere else.
We are not defining any fields, none of this leaves this function.

HenkMutsaerts replied:
We are going to need this somewhere else, since there will be more cases (e.g., is this velocity encoding yes/no, etc.). Hadamard won't be the only exception to PCASL that we will use. So I would move this to a more general parent function (in fact, I believe we usually already have something like `x.iSubject`, `x.SubjectName` etc.).

This would make the code more readable. The priority of study pars is a general thing, the reading of data par is general, etc. Not only for a minor sub-sub-function that checks if we have Hadamard or not.

But if I don't convince you with this, just resolve this conversation :)

jan-petr replied:
Yes, haven't convinced me :D

I mean - I fully agree. In a general case. But here, we need to know this only for a small specific subtask of DICOM volume sorting. So we can only hardly move this elsewhere, because we only need it here. We don't need to know these things in general for other sequences. So this is really a minor sub-sub-function that does this workaround for this difficult hadamard case.

HenkMutsaerts replied:
It is probably already defined somewhere else. And we WILL need it somewhere else, just wait until we start implementing velocity encoded and all other fancy flavors ;)

But OK for now. Please resolve this.

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+%   studyParSpecific - Resolved studyPar for a specific subject/session/run
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION:    Takes studyPar with possibly several contexts and resolves the priority and saves the individual studyPar.
```
That's the whole point - it's a priority list and not inheritance. Therefore, inherently, it will overwrite things and that's the whole point - that things are defined broadly, and then you proceed to specify on session level, visit level or subject and more specific left and overwrite at each time - that's why we have the priorities and that's why it is called like that.

I used the word inheritance while explaining just to make it easier to understand. 

So conflict are dealt with by overwriting by design and there's no point in reporting that. 
Of course we can try to detect clearly duplicity entries like you have, but the point is that you ARE overwriting.

HenkMutsaerts replied:
So inheritance from top to bottom, where each new layer overwrites any conflict with the previous layer?

This is really complicated for the user. Again, the Table ... we need it now. This is asking for trouble.
I'm fine merging this now, on the condition that this doesn't appear in a release without the Table.
Let users (including our PhD students!) use the Table.

Just to explain to you: imagine that you get the same salary throughout the year, but there are a few exceptions in which you get reimbursements, and halfway through the year you get a salary raise. Even though you are right that you can describe this easier, you still want to check on your account that you actually received your salary; your will check this month by month right? ;) That's the same with this case.

So yes, your JSON solution is helpful in a simple case, where each participant has an identical number of sessions and the second session always has a different PLD than the first. But reality is unfortunately different:
A) one of the sessions failed so they repeat it; so a few subjects have 3 sessions
B) most of the subjects have 2 visits where the second visit has a PASL and the first a PCASL; but because the scanner was replaced in the middle, this is not consistent. Some subjects also had PASL on baseline.
And I can go on and on; I have seen this many times. We need the Table. Tables are human readable.

jan-petr replied:
JSON is also easier in this case. Cause you define the general case with low priority and more exceptional cases are redefined when needed. There's no scenario that would have been faster and less error-prone to describe with a table ;)

sorry, but your examples are not good. If checking the salary, I would check - for the whole year, I've received XX EUR, for two months YY EUR.

Just try to imagine how difficult will be to fill in the table. Copy pasting the content 1000 times for 1000 subjects. then trying to find where you made a mistake :) Then you add two more subjects and you need to repeat this whole again :)

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+studyParSpecific = struct();
+
+% At each step, check for all conditions and set a flag for overwrite (first step having the lowest priority
+for iContext = 1:length(studyParFull.StudyPars)
```
Yes - agree. Done

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	return;
+end
+
+% Initialize empty studyPar and then go through different contexts
```
Good point. I forgot to change comments, only the keywords.

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	% Otherwise check if the respective field and Subject/Visit/Session name matches the regular expression
+	
+	% Check subject name - missing rule or missing regexp means that it is OK
+	if isfield(studyParFull.StudyPars{iContext},'SubjectRegexp')
```
I know, but this doesn't really matter. We use very often regexp.

HenkMutsaerts replied:
Well it's wrong. That we do something wrong many times is not an excuse right :) I checked it. It should be RegExp. That's easy to fix right (just for his case).

jan-petr replied:
ok. changing this. 
It's easy - just the code, flavors, documentation :)

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	% Check subject name - missing rule or missing regexp means that it is OK
+	if isfield(studyParFull.StudyPars{iContext},'SubjectRegexp')
+		if ~isempty(subjectName) && ~isempty(studyParFull.StudyPars{iContext}.SubjectRegexp)
+			% If the regexp is not found, then we don't take this subejct
```
Done

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+				bOverwrite = 0;
+			end
+		end
+		studyParFull.StudyPars{iContext} = rmfield(studyParFull.StudyPars{iContext},'SessionRegexp');	
```
Because - all is then passed to the JSON-BIDSify and we risk to write these to the JSON by mistake. It shouldn't happen, but I'd like to keep the code clean and remove things that have nothing to do there.

HenkMutsaerts replied:
OK, but then this should happen only at the end of the code, before it is passed to JSON-BIDSify. Now it limits debugging, because you do this for each subject/session/run level specifically.

jan-petr replied:
I don't see the issue. this function is supposed to take studyParFull and return a clean studyPar with a single instance. It has to be removed inside this function.

HenkMutsaerts replied:
Isn't outputting to a different variable better programming? Because that makes for easier bugfixing. So instead of returning the clean studyPar to the same parameter, you put it in `studyParClean`, and at the end of the iterations you overwrite `x.studyPar` by `studyParClean`.

jan-petr replied:
Sure. Solved by not touching studyPar anymore and cleaning the keywords at once in the final studyParSpecificSubjVisitSess
8ebcd98b35fa2b848d646952c1ce2cd3852d7f86

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+		listFields = fieldnames(studyParFull.StudyPars{iContext});
+		
+		for iField = 1:length(listFields)
+			studyParSpecific.(listFields{iField}) = studyParFull.StudyPars{iContext}.(listFields{iField});
```
Good point. But already written. Next time

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	% Checks for fields SubjectRegexp, VisitRegexp, SessionRegexp - if missing or empty, then take this context into account
+	% Otherwise check if the respective field and Subject/Visit/Session name matches the regular expression
+	
+	% Check subject name - missing rule or missing regexp means that it is OK
```
Good point. Done.

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+for iContext = 1:length(studyParFull.StudyPars)
+	bOverwrite = 1;
+	
+	% Checks for fields SubjectRegexp, VisitRegexp, SessionRegexp - if missing or empty, then take this context into account
```
Yes. I don't see that huge difference between `studyParAll` and `studyParFull`, and `studyParSpecific` and `studyParIndivSubjVisitSess` to justify the change ;)

HenkMutsaerts replied:
OK but then you need to explain in the header what `studyParSpecific` means. Sorry for being annoying, again my mission is to make reading code easier for the non-educated programmer ;)

jan-petr replied:
It is already explained in the function header. Note that `studyParIndivSubjVisitSess` isn't any better or self-explanatory. We can discuss if `indiv` or `specific` is a better keyword though.

HenkMutsaerts replied:
Well `specific`is not self-explanatory. Specific to what? Specific to the labeling? The readout? This vendor?

So I'm also fine with `studyParSpecificSubjVisitSess`. it would be nice to also explain in the subject header such variable names that are not required for input output but are used in the function only.

You can choose and resolve this conversation.

jan-petr replied:
ok - fine with that suggestion. That variable is explained in the header of the called function ;)

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	% Check sessions
+	if isfield(studyParFull.StudyPars{iContext},'VisitRegexp')
+		if ~isempty(sessionName) && ~isempty(studyParFull.StudyPars{iContext}.VisitRegexp)
+			% If the regexp is not found, then we don't take this subejct
```
fixed.

jan-petr commented on Functions/Import/xASL_imp_StudyParPriority.m:
```diff
@@ -0,0 +1,90 @@
+	% Check runs
+	if isfield(studyParFull.StudyPars{iContext},'SessionRegexp')
+		if ~isempty(runName) && ~isempty(studyParFull.StudyPars{iContext}.SessionRegexp)
+			% If the regexp is not found, then we don't take this subejct
```
fixed

jan-petr commented on Functions/InputOutput/xASL_io_ReadDataPar.m:
```diff
@@ -1,12 +1,12 @@
-function [x] = xASL_io_ReadDataPar(pathDataPar,bStudyPar)
+function [x] = xASL_io_ReadDataPar(pathDataPar, bStudyPar)
```
We had 10 functions for this that I have merged to one on a previous request. We can split, but please create a new issue for that.

HenkMutsaerts replied:
I cannot imagine that I would have approved merging 10 different fruits, and then calling them all "apple" ;)
Can you please write the issue? You know more details of what happened last time, and why they are joint.

To me, it doesn't seem sensible to join them. Reading sourceStructure (for dcm2nii), reading studyPar (for nii2BIDS), reading dataPar (for processing) should all be different, if we generalize this we loose modularity and things could go wrong? So I don't see the benefit of why you mixed this in the first place?

Perhaps I can understand historically that we read the dataPar for the parameters that belong to the data, that now are in the JSON sidecars, and `derivatives` and `temp` both have the legacy structure.

This is a remnant from a merged import and processing, we should separate this ;)

jan-petr replied:
We had 2-3 functions for reading JSON with bizarre names. I have merged them to a single one (historically called ReadDataPar), but we can rename it to readJSON or something else.

It's main purpose is to read and fix manually created JSONs = studyPar and dataPar. It's not used for anything else. No sourcestructure json.

The function is exactly the same for studyPar and dataPar and the only difference is that for dataPar, we correct for any errors in data-parameters - which is not needed for studyPar, because there shouldnt' be any dataPars.

That's the obvious benefit of having a single function. And makes no sense to have two functions. If you want, we can remove this option and move this parameter correction outside - I can then create an issue for that or even do it here - up to you.

But it is wrong to regard this as merged import and processing - this has nothing to do with it, it's just a service function.

HenkMutsaerts replied:
OK, I created a new issue for this: #1193

jan-petr commented on Functions/InputOutput/xASL_io_ReadDataPar.m:
```diff
@@ -147,37 +147,56 @@
-			warning('JSON field type could be invalid...');
+	if isfield(x,'StudyPars')
+		bStudyPars = 1;
+		nContext = length(x.StudyPars);
```
fixed

jan-petr commented on Modules/SubModule_Import/xASL_wrp_NII2BIDS.m:
```diff
@@ -48,18 +48,28 @@
 	if ~exist(x.dir.studyPar,'file')
 		warning('Could not find the studyPar.json file...');
-		studyPar = struct;
+		studyParFull = struct;
```
as above...

HenkMutsaerts reviewed (changes requested):
Please provide a warning for each inheritance case, and see my comment for studyParClean, otherwise I'm fine.

jan-petr commented:
> Please provide a warning for each inheritance case, and see my comment for studyParClean, otherwise I'm fine.

OK. I fixed the studyParClean. In a bit different way - I keep studyPar as it was without changing it. And I only clean the fields from the result. This is even more clean solution.

I made a few more edits addressing a few more comments. Can I resolve the rest?

Regarding the warnings - we have to keep it relatively simple. And we can't report on general basis. As we want to keep things modular and this report will change based on what subjects are included and can therefore differ on a re-run with a more complete dataset.

So I propose to report something like this:
`Importing subject XX, session YY, run ZZ. `
`Note: fields AA, BB, CC were updated as multiple studyPars were provided in studyPar.json`

`Importing subject XX, session YY, run ZZ. `
`Exception subject 11|12, session 2|3, run1: Updated fields AA, BB, CC`
`Exception subject 12, session 3, run1: Updated fields AA, BB, CC, DD`

Any preference?

jan-petr commented:
HENK: When it reads the non first study par, it warns "Note that for subjects XX sessions YY we're using parameter XX instead of YY"

And check the main issue comment.

jan-petr commented:
I have copied the same answer to PR as well:
> 
> * We have multiple `studyParameters`, not sure where "context" comes from?
Good catch. That's because we called it ImportContexts previously. I have fixed this everywhere. [501c564](https://github.com/ExploreASL/ExploreASL/pull/1186/commits/501c564f97cc6f6da1dd7530c304c655959eab5b)

> * We create a function that parses `studyPar.json`. This provides a heads-up to the user for each aliasHierarchy, e.g.,:
>   "Note that for cases that match subject regular expression ˆsub-\d{3} session regular expression ... run regular expression, we will change the data parameters XX YY and ZZ.""

This doesn't make sense IMHO. Because this only states the basic functionality of the multi-parameter studyPar :) It doesn't make to report general warning, it makes sense to only report specific warnings to cases that are actually happening. 
So I would go with the warning you previously agreed upon: "When it reads the non-first study par, it warns "Note that for subjects XX sessions YY we're using parameter XX instead of YY"

An example of a current warning is:
`Evaluating multi-parameter studyPar for subject/visit/session: Sub1/1/2`
` For RegExp subject/visit/session //2, overwriting fields:  LabelingDuration, PostLabelingDelay, TimeEncodedMatrixSize, ASLContext,`

> * Later, we will expand this with provenance: [Import provenance #1190](https://github.com/ExploreASL/ExploreASL/issues/1190)
OK. Will check later

I agree in general with #1193, but note that we will anyway revamp the legacy structure of parameters that dataPar uses - so things might still change a bit. Anyway - I agree that we have to clean that. These things happen step-by-step - don't criticize when things are a mess, because they were a super-mess before :D

BeatrizPadrela reviewed (approved):
It works!!

HenkMutsaerts commented:
No I wouldn't provide this information on a subject-by-subject basis, because that will be lost in the long log-list. Instead, at the beginning of the whole conversion: you give a single big warning:

>> Evaluating multi-parameter studyPar for all subjects/visits/sessions:
For subjects 001, 002, 003, visits 002, session 001, we overwrite the following fields: LabelingDuration, PostLabelingDelay, TimeEncodedMatrixSize, ASLContext:
>> * LabelingDuration 1.8 becomes 2.0
>> * PLD 2.0 becomes 1.5
..

jan-petr commented:
> No I wouldn't provide this information on a subject-by-subject basis, because that will be lost in the long log-list. Instead, at the beginning of the whole conversion: you give a single big warning:
> 
> > > Evaluating multi-parameter studyPar for all subjects/visits/sessions:
> > > For subjects 001, 002, 003, visits 002, session 001, we overwrite the following fields: LabelingDuration, PostLabelingDelay, TimeEncodedMatrixSize, ASLContext:
> > > 
> > > * LabelingDuration 1.8 becomes 2.0
> > > * PLD 2.0 becomes 1.5
> > >   ..
> > >   But I'm fine if you move this to the future provenance issue, and close this conversation.

But this is verbatim what is written in the studyPar.json...
